<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A unified framework of realistic in silico data generation and statistical model inference for single-cell and spatial omics • scDesign3</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A unified framework of realistic in silico data generation and statistical model inference for single-cell and spatial omics">
<meta name="description" content="We present a statistical simulator, scDesign3, to generate realistic single-cell and spatial omics data, including various cell states, experimental designs, and feature modalities, by learning interpretable parameters from real data. Using a unified probabilistic model for single-cell and spatial omics data, scDesign3 infers biologically meaningful parameters; assesses the goodness-of-fit of inferred cell clusters, trajectories, and spatial locations; and generates in silico negative and positive controls for benchmarking computational tools.">
<meta property="og:description" content="We present a statistical simulator, scDesign3, to generate realistic single-cell and spatial omics data, including various cell states, experimental designs, and feature modalities, by learning interpretable parameters from real data. Using a unified probabilistic model for single-cell and spatial omics data, scDesign3 infers biologically meaningful parameters; assesses the goodness-of-fit of inferred cell clusters, trajectories, and spatial locations; and generates in silico negative and positive controls for benchmarking computational tools.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">scDesign3</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/scDesign3.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/scDesign3/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level2">
<h2 id="scdesign3">scDesign3<a class="anchor" aria-label="anchor" href="#scdesign3"></a>
</h2>
<hr>
<p>The R package <strong>scDesign3</strong> is an all-in-one single-cell data simulation tool by using reference datasets with different cell states (cell types, trajectories or and spatial coordinates), different modalities (gene expression, chromatin accessibility, protein abundance, DNA methylation, etc), and complex experimental designs. The transparent parameters enable users to alter models as needed; the model evaluation metrics (AIC, BIC) and convenient visualization function help users select models. <span style="color:blue"> <strong>Detailed tutorials that illustrate various functionalities of scDesign3 are available at this <a href="https://songdongyuan1994.github.io/scDesign3/docs/index.html" class="external-link">website</a></strong></span>. The following illustration figure summarizes the usage of scDesign3:</p>
<p><img src="reference/figures/scDesign3_illustration.png" width="600"></p>
<p>To find out more details about <strong>scDesign3</strong>, you can check out our manuscript on Nature Biotechnology:</p>
<p><a href="https://www.nature.com/articles/s41587-023-01772-1" class="external-link">Song, D., Wang, Q., Yan, G. <em>et al.</em> scDesign3 generates realistic in silico data for multimodal single-cell and spatial omics. <em>Nat Biotechnol</em> <strong>42</strong>, 247–252 (2024).</a></p>
<p>The computational time is quadratic to the number of features used in copula modeling. Reducing this number will greatly speed up the calculation.</p>
<p>Please note that the parallel computing of scDesign3 is mainly designed for <strong>UNIX OS</strong>; be careful when you set <code>n_cores</code>. Please note that you should consider <strong>the balance</strong> between <code>n_cores</code> and your ROM (memory). Simply increasing the number of cores without the increase of memory will slow down or froze your program. We recommend that you should allocate at least 1 GB for 1 core.</p>
</div>
<div class="section level1">
<div class="page-header"><h1 id="table-of-contents">Table of contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h1></div>
<ol style="list-style-type: decimal">
<li><a href="#installation-">Installation</a></li>
<li><a href="#quick-start">Quick Start</a></li>
<li><a href="#tutorials">Tutorials</a></li>
<li><a href="#contact">Contact</a></li>
<li><a href="#related-manuscripts">Related Manuscripts</a></li>
</ol>
<div class="section level2">
<h2 id="installation">Installation<a name="installation-"></a>
<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the development version from GitHub, please run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"SONGDONGYUAN1994/scDesign3"</span><span class="op">)</span></span></code></pre></div>
<p>We are now working on submitting it to Bioconductor and will provide the link once online.</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a name="quick-start"></a>
<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The following code is a quick example of running our simulator. The function <code><a href="reference/scdesign3.html">scdesign3()</a></code> takes in a <code>SinglecellExperiment</code> object with the cell covariates(such as cell types, pseudotime, or spatial coordinates) stored in the <code>colData</code> of the <code>SinglecellExperiment</code> object. For more details on the <code>SinlgeCellExperiment</code> object, please check on its <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">Bioconductor link</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_simu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/scdesign3.html">scdesign3</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    celltype <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    pseudotime <span class="op">=</span> <span class="st">"pseudotime"</span>,</span>
<span>    spatial <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    other_covariates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    mu_formula <span class="op">=</span> <span class="st">"s(pseudotime, k = 10, bs = 'cr')"</span>,</span>
<span>    sigma_formula <span class="op">=</span> <span class="st">"s(pseudotime, k = 5, bs = 'cr')"</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    correlation_function <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>    usebam <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    corr_formula <span class="op">=</span> <span class="st">"1"</span>,</span>
<span>    copula <span class="op">=</span> <span class="st">"gaussian"</span>,</span>
<span>    fastmvn <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    DT <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    pseudo_obs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    family_set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gauss"</span>, <span class="st">"indep"</span><span class="op">)</span>,</span>
<span>    important_feature <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>    nonnegative <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    return_model <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    nonzerovar <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"mcmapply"</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    trace <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>The parameters of <code><a href="reference/scdesign3.html">scdesign3()</a></code> are:</p>
<ul>
<li>
<code>sce</code>: A SingleCellExperiment object.</li>
<li>
<code>assay_use</code>: A string which indicates the assay you will use in the sce. Default is ‘counts’.</li>
<li>
<code>celltype</code>: A string of the column name of the cell type variable in the colData of the sce. Default is ‘cell_type’. The cell type variable in the colData of the sce should be a factor variable. Use NULL if there is no column in the colData that contains the cell-type information.</li>
<li>
<code>pseudotime</code>: A string or a string vector of the name of pseudotime and (if exist) multiple lineages. Default is NULL. If the data only has one lineage, then this parameter should be the column name of the pseudotime variable in the colData of the sce. If the data has multiple lingaes, then this parameter be the column names of the pseudotime variables for each lineage and the variables indicating which lineage that a cell belongs to. The pseudotime variables should be continuous numeric variables.</li>
<li>
<code>spatial</code>: A length-two string vector of the column names of spatial coordinates in the colData of sce. Default is NULL.</li>
<li>
<code>other_covariate</code>: A string or a string vector of the other covariates in the colData of sce you want to include in the data. For example, you can put the column names of the batch variables and/or condition variables in the colData of sce here if your sce contains these information and you want to include these variables in<code>mu_formula</code> or <code>sigma_formula</code> or <code>corr_formula</code>.</li>
<li>
<code>mu_formula</code>: A string of the mu parameter formula for fitting each gene’s marginal distribution.</li>
<li>
<code>sigma_formula</code>: A string of the sigma parameter formula for fitting each gene’s marginal distribution.</li>
<li>
<code>family_use</code>: A string of the marginal distribution you want to use when fitting each gene’s marginal distribution. Must be one of ‘poisson’, ‘nb’, ‘zip’, ‘zinb’ or ‘gaussian’.</li>
<li>
<code>n_cores</code>: An integer. The number of cores to use.</li>
<li>
<code>correlation_function</code>: A string. If ‘default’, the function from Rfast; if ‘coop’, the function from coop, which calls BLAS.</li>
<li>
<code>usebam</code>: A logic variable. If TRUE, use bam (generalized additive models for very large datasets) for acceleration.</li>
<li>
<code>edf_flexible</code>: A logic variable. If TRUE, the degree of freedom for each gene’s regression model will be automatically selected for acceleration.</li>
<li>
<code>corr_formula</code>: A string of the correlation structure. For example, if you want to obtain a correlation structure for each cell type, then this parameter should be the column name of the cell type variable in the colData of sce.</li>
<li>
<code>copula</code>: A string of the copula choice. Must be one of ‘gaussian’ or ‘vine’. Default is ‘gaussian’. Note that vine copula may have better modeling of high-dimensions, but can be very slow when features are &gt;1000.</li>
<li>
<code>fastmvn</code>: An logical variable. If TRUE, the sampling of multivariate Gaussian is done by mvnfast, otherwise by mvtnorm. Default is FALSE. It only matters for Gaussian copula.</li>
<li>
<code>DT</code>: A logic variable. If TRUE, perform the distributional transformation to make the discrete data ‘continuous’. This is useful for discrete distributions (e.g., Poisson, NB). Default is TRUE. Note that for continuous data (e.g., Gaussian), DT does not make sense and should be set as FALSE.</li>
<li>
<code>pseudo_obs</code>: A logic variable. If TRUE, use the empirical quantiles instead of theoretical quantiles for fitting copula. Default is FALSE.</li>
<li>
<code>family_set</code>: A string or a string vector of the bivariate copula families. Default is c(“gauss”, “indep”).</li>
<li>
<code>important_feature</code>: A string or vector which indicates whether a gene will be used in correlation estimation or not. If this is a string, then this string must be “all” or “auto”, which indicates that all genes are used or the genes will be automatically selected based on the proportion of zero expression across cells for each gene. Gene with zero proportion greater than 0.8 will be excluded form gene-gene correlation estimation. If this is a vector, then this should be a logical vector with length equal to the number of genes in sce. TRUE in the logical vector means the corresponding gene will be included in gene-gene correlation estimation and FALSE in the logical vector means the corresponding gene will be excluded from the gene-gene correlation estimation. The default value is “all”.</li>
<li>
<code>nonnegative</code>: A logical variable. If TRUE, values &lt; 0 in the synthetic data will be converted to 0. Default is TRUE (since the expression matrix is nonnegative).</li>
<li>
<code>return_model</code>: A logic variable. If TRUE, the marginal models and copula models will be returned. Default is FALSE.</li>
<li>
<code>nonzerovar</code>: A logical variable. If TRUE, for any gene with zero variance, a cell will be replaced with 1. This is designed for avoiding potential errors, for example, PCA.</li>
<li>
<code>parallelization</code>: A string indicating the specific parallelization function to use. Must be one of ‘mcmapply’, ‘bpmapply’, or ‘pbmcmapply’, which corresponds to the parallelization function in the package parallel,BiocParallel, and pbmcapply respectively. The default value is ‘mcmapply’.</li>
<li>
<code>BPPARAM</code>: A MulticoreParam object or NULL. When the parameter parallelization = ‘mcmapply’ or ‘pbmcmapply’, this parameter must be NULL. When the parameter parallelization = ‘bpmapply’, this parameter must be one of the MulticoreParam object offered by the package ’BiocParallel. The default value is NULL.</li>
<li>
<code>TRACE</code>: A logic variable. If TRUE, the warning/error log and runtime for gam/gamlss will be returned, FALSE otherwise. Default is FALSE.</li>
</ul>
<p>The output of <code><a href="reference/scdesign3.html">scdesign3()</a></code> is a list which includes:</p>
<ul>
<li>
<code>new_count</code>: This is the synthetic count matrix generated by <code><a href="reference/scdesign3.html">scdesign3()</a></code>.</li>
<li>
<code>new_covariate</code>:
<ul>
<li>If the parameter <code>ncell</code> is set to a number that is different from the number of cells in the input data, this will be a matrix that has the new cell covariates that are used for generating new data.</li>
<li>If the parameter <code>ncell</code> is the default value, this will be <code>NULL</code>.</li>
</ul>
</li>
<li>
<code>model_aic</code>: This is a vector include the genes’ marginal models’ AIC, fitted copula’s AIC, and total AIC, which is the sum of the previous two.</li>
<li>
<code>model_bic</code>: This is a vector include the genes’ marginal models’ BIC, fitted copula’s BIC, and total BIC, which is the sum of the previous two.</li>
<li>
<code>marginal_list</code>:
<ul>
<li>If the parameter <code>return_model</code> is set to <code>TRUE</code>, this will be a list which contains the fitted gam or gamlss model for all genes in the input data. This may greatly increase the object size.</li>
<li>If the parameter <code>return_model</code> is set to the default value <code>FALSE</code>, this will be <code>NULL</code>.</li>
</ul>
</li>
<li>
<code>corr_list</code>:
<ul>
<li>If the parameter <code>return_model</code> is set to <code>TRUE</code>, this will be a list which contains either a correlation matrix (when <code>copula = "gaussian"</code>) or the fitted Vine copula (when <code>copula = "vine"</code>) for each user specified correlation groups (based on the parameter <code>corr_by</code>).</li>
<li>If the parameter <code>return_model</code> is set to the default value <code>FALSE</code>, this will be <code>NULL</code>.</li>
</ul>
</li>
</ul>
<p>For more details about the <code>mu_formula</code> and <code>sigma_formula</code> formula specification, please check online materials about the package <a href="https://cran.r-project.org/web/packages/mgcv/index.html" class="external-link">mgcv</a>. Technically speaking, you can try any formulas as long as they are available for <strong>mgcv</strong>.</p>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a name="tutorials"></a>
<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<p>For all detailed tutorials, please check the <a href="https://songdongyuan1994.github.io/scDesign3/docs/index.html" class="external-link">website</a>. The tutorials will demonstrate the applications of <strong>scDesign3</strong> from the following four perspectives: data simulation, model parameters, model selection, and model alteration.</p>
<ul>
<li>Data simulation
<ul>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-cellType-vignette.html" class="external-link">Simulate datasets with cell type and modified cell-type proportions</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-librarySize-vignette.html" class="external-link">Simulate datasets with cell library size</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-multipleLineages-vignette.html" class="external-link">Simulate datasets with multiple lineages</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-spatial-vignette.html" class="external-link">Simulate spatial transcriptomic data</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-spatial-deconvolution.html" class="external-link">Simulate spot-resolution spatial data for cell-type deconvolution</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-scATACseq-vignette.html" class="external-link">Simulate single-cell ATAC-seq data</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-CITEseq-vignette.html" class="external-link">Simulate CITE-seq data</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-multiomics-vignette.html" class="external-link">Simulate multi-omics data from multiple single-omic datasets</a></li>
</ul>
</li>
<li>Model parameter
<ul>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-introduction-vignette.html" class="external-link">scDesign3 introduction</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-marginal-vignette.html" class="external-link">scDesign3 marginal distribution for genes</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-copulaCompare-vignette.html" class="external-link">Compare Gaussian copula and Vine copula</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-parallelization-vignette.html" class="external-link">Parallelization in scDesign3</a></li>
</ul>
</li>
<li>Model selection
<ul>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-clusterGOF-vignette.html" class="external-link">Evaluate clustering goodness-of-fit by scDesign3</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-pseudotimeGOF-vignette.html" class="external-link">Evaluate pseudotime goodness-of-fit by scDesign3</a></li>
</ul>
</li>
<li>Model alteration
<ul>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-batchEffect-vignette.html" class="external-link">Simulate datasets with/without batch effect</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-conditionEffect-vignette.html" class="external-link">Simulate datasets with/without condition effect</a></li>
<li><a href="https://songdongyuan1994.github.io/scDesign3/docs/articles/scDesign3-DEanalysis-vignette.html" class="external-link">Simulate datasets for DE test</a></li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="contact">Contact<a name="contact"></a>
<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>Any questions or suggestions on <code>scDesign3</code> are welcomed! Please report it on <a href="https://github.com/SONGDONGYUAN1994/scDesign3/issues" class="external-link">issues</a>, or contact Dongyuan Song (<a href="mailto:dongyuansong@ucla.edu" class="email">dongyuansong@ucla.edu</a>) or Qingyang Wang (<a href="mailto:qw802@g.ucla.edu" class="email">qw802@g.ucla.edu</a>).</p>
</div>
<div class="section level2">
<h2 id="changelog">Changelog<a class="anchor" aria-label="anchor" href="#changelog"></a>
</h2>
<ul>
<li>2024-11-07
<ul>
<li>Add options for correlation estimation. The alternative is from R package <a href="https://cran.rstudio.com/web/packages/coop/index.html" class="external-link">coop</a>, which requires BLAS</li>
</ul>
</li>
<li>2024-06-22 Important changes
<ul>
<li>Add sparse Gaussian copula for <code>fit_copula</code>
</li>
<li>Add automatic k selection for <code>fit_marginal</code>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="related-manuscripts">Related Manuscripts<a name="related-manuscripts"></a>
<a class="anchor" aria-label="anchor" href="#related-manuscripts"></a>
</h2>
<ul>
<li>The original <strong>scDesign3</strong> paper
<ul>
<li>
<strong>scDesign3</strong>: <a href="https://www.nature.com/articles/s41587-023-01772-1" class="external-link">Song, D., Wang, Q., Yan, G. <em>et al.</em> scDesign3 generates realistic in silico data for multimodal single-cell and spatial omics. <em>Nat Biotechnol</em> <strong>42</strong>, 247–252 (2024).</a>
</li>
</ul>
</li>
<li>The predecessors of <strong>scDesign3</strong>
<ul>
<li>
<strong>scDesign</strong>: <a href="https://academic.oup.com/bioinformatics/article/35/14/i41/5529133" class="external-link">Li, W. V., &amp; Li, J. J. (2019). A statistical simulator scDesign for rational scRNA-seq experimental design. <em>Bioinformatics</em>, <strong>35</strong>(14), i41-i50.</a>
</li>
<li>
<strong>scDesign2</strong>: <a href="https://link.springer.com/article/10.1186/s13059-021-02367-2" class="external-link">Sun, T., Song, D., Li, W. V., &amp; Li, J. J. (2021). scDesign2: a transparent simulator that generates high-fidelity single-cell gene expression count data with gene correlations captured. <em>Genome biology</em>, <strong>22</strong>(1), 1-37.</a>
</li>
</ul>
</li>
<li>The simulator for single-cell multi-omics reads developed by our lab member Guanao Yan
<ul>
<li>
<strong>scReadSim</strong>: <a href="https://doi.org/10.1038/s41467-023-43162-w" class="external-link">Yan, G., Song, D. &amp; Li, J.J. scReadSim: a single-cell RNA-seq and ATAC-seq read simulator. <em>Nat Commun</em> <strong>14</strong>, 7482 (2023)</a>
</li>
</ul>
</li>
</ul>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/scDesign3" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/SONGDONGYUAN1994/scDesign3/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/SONGDONGYUAN1994/scDesign3/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scDesign3</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Dongyuan Song <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1114-1215" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Qingyang Wang <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-1051-609X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Chenxin Jiang <br><small class="roles"> Author </small> <a href="https://orcid.org/0009-0005-7369-4116" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dongyuan Song, Qingyang Wang, Chenxin Jiang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
