<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Fit the marginal models — fit_marginal • scDesign3</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Fit the marginal models — fit_marginal"><meta name="description" content="fit_marginal fits the per-feature regression models."><meta property="og:description" content="fit_marginal fits the per-feature regression models."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDesign3</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scDesign3.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/scDesign3/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Fit the marginal models</h1>
      <small class="dont-index">Source: <a href="https://github.com/SONGDONGYUAN1994/scDesign3/blob/main/R/fit_marginal.R" class="external-link"><code>R/fit_marginal.R</code></a></small>
      <div class="d-none name"><code>fit_marginal.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>fit_marginal</code> fits the per-feature regression models.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fit_marginal</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  predictor <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  <span class="va">mu_formula</span>,</span>
<span>  <span class="va">sigma_formula</span>,</span>
<span>  <span class="va">family_use</span>,</span>
<span>  <span class="va">n_cores</span>,</span>
<span>  usebam <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  edf_flexible <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  parallelization <span class="op">=</span> <span class="st">"mcmapply"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trace <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  simplify <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filter_cells <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>An object from <code><a href="construct_data.html">construct_data</a></code>.</p></dd>


<dt id="arg-predictor">predictor<a class="anchor" aria-label="anchor" href="#arg-predictor"></a></dt>
<dd><p>A string of the predictor for the gam/gamlss model. Default is "gene". This is just a name.</p></dd>


<dt id="arg-mu-formula">mu_formula<a class="anchor" aria-label="anchor" href="#arg-mu-formula"></a></dt>
<dd><p>A string of the mu parameter formula. It follows the format of formula in <code><a href="https://rdrr.io/pkg/mgcv/man/bam.html" class="external-link">bam</a></code>. Note: if the formula has multiple smoothers (<code>s()</code>) (we do not recommend this), please put the one with largest k (most complex one) as the first one.</p></dd>


<dt id="arg-sigma-formula">sigma_formula<a class="anchor" aria-label="anchor" href="#arg-sigma-formula"></a></dt>
<dd><p>A string of the sigma parameter formula</p></dd>


<dt id="arg-family-use">family_use<a class="anchor" aria-label="anchor" href="#arg-family-use"></a></dt>
<dd><p>A string or a vector of strings of the marginal distribution.
Must be one of 'binomial', 'poisson', 'nb', 'zip', 'zinb' or 'gaussian', which represent 'poisson distribution',
'negative binomial distribution', 'zero-inflated poisson distribution', 'zero-inflated negative binomial distribution',
and 'gaussian distribution' respectively.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>An integer. The number of cores to use.</p></dd>


<dt id="arg-usebam">usebam<a class="anchor" aria-label="anchor" href="#arg-usebam"></a></dt>
<dd><p>A logic variable. If use <code><a href="https://rdrr.io/pkg/mgcv/man/bam.html" class="external-link">bam</a></code> for acceleration.</p></dd>


<dt id="arg-edf-flexible">edf_flexible<a class="anchor" aria-label="anchor" href="#arg-edf-flexible"></a></dt>
<dd><p>A logic variable. It uses simpler model to accelerate the marginal fitting with a mild loss of accuracy. If TRUE, the fitted regression model will use the fitted relationship between Gini coefficient and the effective degrees of freedom on a random selected gene sets. Default is FALSE.</p></dd>


<dt id="arg-parallelization">parallelization<a class="anchor" aria-label="anchor" href="#arg-parallelization"></a></dt>
<dd><p>A string indicating the specific parallelization function to use.
Must be one of 'mcmapply', 'bpmapply', or 'pbmcmapply', which corresponds to the parallelization function in the package
<code>parallel</code>,<code>BiocParallel</code>, and <code>pbmcapply</code> respectively. The default value is 'mcmapply'.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A <code>MulticoreParam</code> object or NULL. When the parameter parallelization = 'mcmapply' or 'pbmcmapply',
this parameter must be NULL. When the parameter parallelization = 'bpmapply',  this parameter must be one of the
<code>MulticoreParam</code> object offered by the package 'BiocParallel. The default value is NULL.</p></dd>


<dt id="arg-trace">trace<a class="anchor" aria-label="anchor" href="#arg-trace"></a></dt>
<dd><p>A logic variable. If TRUE, the warning/error log and runtime for gam/gamlss will be returned.
will be returned, FALSE otherwise. Default is FALSE.</p></dd>


<dt id="arg-simplify">simplify<a class="anchor" aria-label="anchor" href="#arg-simplify"></a></dt>
<dd><p>A logic variable. If TRUE, the fitted regression model will only keep the essential contains for <code>predict</code>. Default is FALSE.</p></dd>


<dt id="arg-filter-cells">filter_cells<a class="anchor" aria-label="anchor" href="#arg-filter-cells"></a></dt>
<dd><p>A logic variable. If TRUE, when all covariates used for fitting the GAM/GAMLSS model are categorical, the code will check each unique combination of categories and remove cells in that category if it has all zero gene expression for each fitted gene.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of fitted regression models. The length is equal to the total feature number.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function takes the result from <code><a href="construct_data.html">construct_data</a></code> as the input,
and fit the regression models for each feature based on users' specification.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="construct_data.html">construct_data</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  sce <span class="op">=</span> <span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>  assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>  celltype <span class="op">=</span> <span class="st">"cell_type"</span>,</span></span>
<span class="r-in"><span>  pseudotime <span class="op">=</span> <span class="st">"pseudotime"</span>,</span></span>
<span class="r-in"><span>  spatial <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  other_covariates <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  corr_by <span class="op">=</span> <span class="st">"1"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">my_marginal</span> <span class="op">&lt;-</span> <span class="fu">fit_marginal</span><span class="op">(</span></span></span>
<span class="r-in"><span>  data <span class="op">=</span> <span class="va">my_data</span>,</span></span>
<span class="r-in"><span>  mu_formula <span class="op">=</span> <span class="st">"s(pseudotime, bs = 'cr', k = 10)"</span>,</span></span>
<span class="r-in"><span>  sigma_formula <span class="op">=</span> <span class="st">"1"</span>,</span></span>
<span class="r-in"><span>  family_use <span class="op">=</span> <span class="st">"nb"</span>,</span></span>
<span class="r-in"><span>  n_cores <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  usebam <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dongyuan Song, Qingyang Wang, Chenxin Jiang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

